solr/core/src/java/org/apache/solr/update/AddUpdateCommand.java

```diff
@@ -175,69 +172,83 @@ public String getHashableId() {
-          if(isVersion) sdoc.setField(CommonParams.VERSION_FIELD, version);
-          // TODO: if possible concurrent modification exception (if SolrInputDocument not cloned and is being forwarded to replicas)
-          // then we could add this field to the generated lucene document instead.
-        }
-
-        iter = all.iterator();
-     }
+  public List<SolrInputDocument> computeFlattenedDocs() {
+    List<SolrInputDocument> all = flatten(solrDoc);
 
-      @Override
-      public boolean hasNext() {
-        return iter.hasNext();
-      }
+    String rootId = getHashableId();
 
-      @Override
-      public Document next() {
-        return DocumentBuilder.toDocument(iter.next(), req.getSchema());
-      }
+    boolean isVersion = version != 0;
 
-      @Override
-      public void remove() {
-        throw new UnsupportedOperationException();
+    for (SolrInputDocument sdoc : all) {
+      if (all.size() > 1) {
+        sdoc.setField(IndexSchema.ROOT_FIELD_NAME, rootId);
       }
-    };
+      if(isVersion) sdoc.setField(CommonParams.VERSION_FIELD, version);
+      // TODO: if possible concurrent modification exception (if SolrInputDocument not cloned and is being forwarded to replicas)
+      // then we could add this field to the generated lucene document instead.
+    }
+    return all;
   }
 
   private List<SolrInputDocument> flatten(SolrInputDocument root) {
     List<SolrInputDocument> unwrappedDocs = new ArrayList<>();
-    recUnwrapp(unwrappedDocs, root);
+    recUnwrapAnonymous(unwrappedDocs, root, true);
+    recUnwrapRelations(unwrappedDocs, root, true);
     if (1 < unwrappedDocs.size() && ! req.getSchema().isUsableForChildDocs()) {
       throw new SolrException
         (SolrException.ErrorCode.BAD_REQUEST, "Unable to index docs with children: the schema must " +
          "include definitions for both a uniqueKey field and the '" + IndexSchema.ROOT_FIELD_NAME +
          "' field, using the exact same fieldType");
     }
+    unwrappedDocs.add(root);
     return unwrappedDocs;
   }
 
-  private void recUnwrapp(List<SolrInputDocument> unwrappedDocs, SolrInputDocument currentDoc) {
+  /** Extract all child documents from parent that are saved in keys. */
+  private void recUnwrapRelations(List<SolrInputDocument> unwrappedDocs, SolrInputDocument currentDoc, boolean isRoot) {
+    for (SolrInputField field: currentDoc.values()) {
+      Object value = field.getFirstValue();
+      // check if value is a childDocument
+      if (value instanceof SolrInputDocument) {
+        Object val = field.getValue();
+        if (!(val instanceof Collection)) {
+          recUnwrapRelations(unwrappedDocs, ((SolrInputDocument) val));
+          continue;
+        }
+        Collection<SolrInputDocument> childrenList = ((Collection) val);
+        for (SolrInputDocument child : childrenList) {
+          recUnwrapRelations(unwrappedDocs, child);
+        }
+      }
+    }
+
```

Comment:
maybe add a check that these field keyed relations have no anonymous children?  It's quick to check.  The reverse shouldn't be allowed either but I wouldn't bother enforcing that as it's a bit more work.

Reply:
Yeah guess an exception should be thrown

Reply:
I just thought this through, and perhaps it would be better to throw the exception in dataloader e.g. JsonLoader, XmlLoader?

Reply:
but it's not specific to the input type?

Reply:
True, never mind.